### ISSUE TITLE
"npm install" removes nearly all other dependencies

### ISSUE THREAD

**Issue by itinance** [npm/npm#17929]
When running "npm install" nearly all dependencies got removed. This happens on npm 5.3.0. There was formerly an issue opened and already closed, but the issue persists, also for others:
1. https://github.com/npm/npm/issues/16839
2. https://github.com/npm/npm/issues/16839#issuecomment-317849594

I have several reproducable projects where i can replay it (means: all my current active projects).

Example:
I go into a running app with multiple dependencies (react-native). Now i install a new library:

**1. Before "npm install":**

```
$ du -sh node_modules/
128M	node_modules/
```

```
$ ls -l node_modules/react-native
total 88
-rw-r--r--   1 hagen  wheel  1533 Jul 17 23:00 LICENSE
drwxr-xr-x  45 hagen  wheel  1598 Jul 26 00:22 Libraries
-rw-r--r--   1 hagen  wheel  1989 Jul 17 23:00 PATENTS
drwxr-xr-x  13 hagen  wheel   476 Jul 26 00:22 React
-rw-r--r--   1 hagen  wheel  7719 Jul 17 23:00 React.podspec
drwxr-xr-x   4 hagen  wheel   374 Jul 26 00:22 ReactAndroid
drwxr-xr-x   6 hagen  wheel   238 Jul 26 00:22 ReactCommon
drwxr-xr-x   3 hagen  wheel   102 Jul 26 00:22 android
-rw-r--r--   1 hagen  wheel   372 Jul 17 23:00 cli.js
drwxr-xr-x   2 hagen  wheel   442 Jul 26 00:22 flow
drwxr-xr-x   2 hagen  wheel   170 Jul 26 00:22 jest
-rw-r--r--   1 hagen  wheel   696 Jul 17 23:00 jest-preset.json
drwxr-xr-x   2 hagen  wheel   408 Jul 26 00:22 lib
drwxr-xr-x  20 hagen  wheel   918 Jul 26 00:22 local-cli
drwxr-xr-x  12 hagen  wheel   408 Jul 26 00:22 node_modules
-rw-r--r--   1 hagen  wheel  7738 Jul 17 23:00 package.json
-rw-r--r--   1 hagen  wheel  5048 Jul 17 23:00 react.gradle
drwxr-xr-x   2 hagen  wheel   272 Jul 26 00:22 scripts
-rw-r--r--   1 hagen  wheel  1508 Jul 17 23:00 setupBabel.js
drwxr-xr-x   2 hagen  wheel   170 Jul 26 00:22 third-party-podspecs
```

**2. "npm install":**

```
$ npm install react-native-video-processing --save
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/jest as it wasn't installed by /data/workspace/cbl/video_local/node_modules/jest
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/uglifyjs as it wasn't installed by /data/workspace/cbl/video_local/node_modules/uglify-js
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/sane as it wasn't installed by /data/workspace/clb/video_local/node_modules/sane
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/uuid as it wasn't installed by /data/workspace/cbl/video_local/node_modules/uuid
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/json5 as it wasn't installed by /data/workspace/cbl/video_local/node_modules/json5
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/esparse as it wasn't installed by /data/workspace/cbl/video_local/node_modules/esprima
npm WARN gentlyRm not removing /data/workspace/cbl/video_local/node_modules/.bin/esvalidate as it wasn't installed by /data/workspace/Cooblr/video_local/node_modules/esprima
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN react-native-video-processing@1.12.5 requires a peer of react-native@>=0.35 but none was installed.

+ react-native-video-processing@1.12.5
added 26 packages, removed 585 packages and updated 82 packages in 14.333s
```

The disk usage of "node_modules" is after this command only 33 MB. Formerly it was up to 120 MB.

```
$ du -sh node_modules/
 33M	node_modules/

```

"npm install" removed nearly all dependencies.

The removed dependencies are looking like this now:

```
$ ls -lr node_modules/react-native
total 0
drwxr-xr-x  3 hagen  wheel  102 Jul 26 00:16 node_modules
```

**Versions:**
$ npm -v && node -v
5.3.0
v6.2.2

I can replay this all the time.

**Comment by ianvonholt** [comment]
Happening with private repositories as well: #17927

**Comment by davidsturgeon** [comment]
Also getting this issue, when I npm install its removing packages.
added 13 packages, removed 365 packages and updated 3 packages in 29.673s

**Comment by mannyyang** [comment]
same here. getting package removals when installing a dependency.
added 11 packages and removed 458 packages in 54.176s

**Comment by ruchernchong** [comment]
I am having this issue and it has been a problem since the release of `npm v5`. I am temporarily on `yarn` right now because of this. It does not make sense to keep deleting `node_modules` folder and `package-lock.json` on production just to circumvent this problem.

Also, the lock file keeps getting modified when running `npm install` on production.

**Comment by iyobo** [comment]
@ruchern That's the bad part. My production builds are no longer idempotent.
Might need to just downgrade to npm4 at this rate while all this gets sorted out.

**Comment by ruchernchong** [comment]
@iyobo Perhaps you might want to try using `yarn` until this problem gets sorted out?

**Comment by iyobo** [comment]
@ruchern Thing is, I use a mono-repo tool called Lerna to manage my workspace and that uses npm.
Right now, it's like every deploy is immediately followed by sessions of going from server to server re-installing modules recursively through the node_module of modules in node_modules. :(

It's pretty nasty.
I know npm4 was slow, but at least it works for now :)

**Comment by ruchernchong** [comment]
Have you actually tried using `yarn` yet? AFAIK, `yarn` and `npm` works the same way; they are just different in some ways.

Anyway, I should not be talking about `yarn` here, but for now, my temporary solution is that for those who might want to follow.

**Comment by JeremyCraigMartinez** [comment]
I get this behavior only when I pass in `--save`. But when I just run (for example) `npm i jsonschema` (without `--save`) it writes to my package.json anyway and does not remove the rest of the dependencies.

So I guess a temporary solution would be to stop passing the `--save` flag. And if you are installing a package just temporarily to test it out, be mindful to run `npm uninstall jsonschema` before staging and committing you're `package.json`/`package-lock.json` files. But this behavior should be corrected to not remove all of your dependencies..

**Comment by JeremyCraigMartinez** [comment]
Also, `--save-dev` works fine it seems

**Comment by marcoscaceres** [comment]
@JeremyCraigMartinez it happens irrespective of what one passes as an argument - particularly when `package.json` has both dev and non-dev dependency members (and one really needs `npm install`). Having to install all packages with `--save-dev` wouldn't really scale. 

Best solution I've personally found is to just downgrade with nvm: `nvm use v7.10`. Hopefully this gets fixed. It's been happening since the 5.0 release.

**Comment by pkmuz1** [comment]
please help me i can't able to install npm in kali linux

when i hit command - npm install
                                  bash: npm: command not found

please help if you can

**Comment by jbcpollak** [comment]
@pkmuz1 - this is a different issue. For your problem please look into how
to install NodeJS for your Linux version.

**Comment by Butterfly** [comment]
@pkmuz1, in addition to @jbcpollak recommendation, did you close your terminal window and then reopen it? I'm brand new at this so that tip is fresh in mind right now. It works.. frequently. :smile:

If closing the terminal doesn't help, Node.js' [Installing via package manager page](https://nodejs.org/en/download/package-manager/) has a curl command line that I think helped in *my* case.  Node.js' suggestion is about installing nodejs, TOO, so maybe that's not a universally appropriate suggestion for that reason. Feels right in this case, though, because that's what I'm understanding @jbcpollak to be telling you that you should try.

A side note about npm within nodejs: The glitch in my case is that Debian Linux Buster (testing, not stable) currently does NOT show npm as findable and thus installable as a standalone. As of yesterday, npm was also NOT found as a file within nodejs for Buster (yes, it was found there for other Debian releases). [Following Node.js' curl related suggestion](https://nodejs.org/en/download/package-manager/) there appears to have helped install a potentially functioning npm in spite of that. :smile:

**Comment by Butterfly** [comment]
@itinance For however it matters, I'm using a debootstrap'ed (bare minimum) Debian Linux Buster (testing, sits in between stable and unstable).

I had written a reaction to your troubles, and then suddenly it hit me that you're writing about what I think I just went through... over and over and *OVER* again with peer dependencies. It was a mind blowing vicious circle where I would install a specifically declared [version]... and then... when that list of remaining unsolved peer dependencies came up in the terminal window... packages I had JUST installed five minutes before were BACK needing to be installed AGAIN..

That happened AT LEAST 6 or 8 times before it cognitively hit my brain that something wasn't working right. It was so bad that I was considering nuking my entire very first node related effort and starting over...

But then I remembered someone somewhere saying... deleting node_modules, doing things like that momentarily help keep a personal project moving forward... but it doesn't help permanently solve the underlying glitch for everyone....

In my case, the npm install offenders were eslint and webpack. Seriously... AT LEAST 6 or 8 installs later, that list of unsolved peer dependencies would ebb and flow in size with those two being the only remaining unsolved items on that list. I am *so* glad to just now realize it wasn't just me as a newbie flubbing something up... :smile:

**Comment by ruchernchong** [comment]
Hi @zkat, are there any updates on this issue?

**Comment by Butterfly** [comment]
An immediate afterthought... Could this have anything to do with what flags, e.g. "--save" and "--save-dev", that we use on our dependencies? Rightly or wrongly, my current understanding is that at least some of those "just" throw references into lumps, but maybe there is some implication toward their weight or implied stability in all this?

Being BRAND new at this (and also very bold), I have no problem saying that I currently have not one clue about the implications of those "lumps". I am instead wholly at the mercy of whatever tipsters bubble to the top of Internet searches related to *just* getting started with installations in the magical world of nodes. :smile:

**Comment by Butterfly** [comment]
Ok, I have one more observation. I have now better grasped the #16938 [npm install produces different lockfiles on different computes](https://github.com/npm/npm/issues/16938) bug. Most specifically is [the comment by Raigen](https://github.com/npm/npm/issues/16938#issuecomment-322705234): 

> Like this I am not able to contribute new dependencies to the project because it will not just add new packages to packages-lock.json but also changes all integrity fields in the whole file. 

Yes, I know it's not exactly the same, but the uninformed newbie in me sees a potential rationale for where dependencies could be repeatedly replacing each other as they did in my case.

My impression during my experience was that maybe some type of logging file kept being changed that shouldn't have. That would at least partially explain why my remaining two packages, eslint and webpack, went in endless circles deleting their own respective "siblings" in both directions. They both deleted their respective same named packages numbered both higher AND lower than any targeted install version of themselves.

What this does *not* explain is why this did *not* happen to *all* the peer dependencies encountered (~48 of them), It only happened with those two, eslint and webpack. No, it was not just at the end or something, either. This bug occurred throughout the entire peer dependency installation process during which I repeatedly tested them hoping to see something change for the better. :smile:

**Comment by schmod** [comment]
Does this happen if you run `npm install` before running `npm install --save pkg-name`?  

I'm finding it curious that you have a `node_modules` folder, but no `package-lock.json` file, unless you came directly from npm v4.

**Comment by zkat** [comment]
isn't this just because you're generating a `node_modules` using yarn?

npm is going to delete anything it didn't install itself.

**Comment by mcollina** [comment]
@zkat npm removed eslint from a repo today. npm complained that it was not satisfying the peer dependencies, so I knew there was something wrong.

I typically switch between node and npm versions a lot, so it might be related. I am not a package-lock user (mainly for the reason above).

**Comment by ianvonholt** [comment]
@zkat In my case, and I believe the original issue submitters case as well, I'm not using yarn at all. Purely npm.

**Comment by marcoscaceres** [comment]
Same here, not using yarn. I'm trying again with deleting package-lock file. It was generated with 5.0, which is when this started happening for me.

**Comment by marcoscaceres** [comment]
Ok, that worked for me... possible work around:

1. Delete package-lock.json. 
2. Delete `node_modules` folder. 
3. `npm install`

**Comment by ianvonholt** [comment]
@marcoscaceres While this indeed does work, and dependencies are properly resolved, this does nothing to alleviate the problem of updating your `package.json` and just running `npm install` again. 

Personally, there are many scenarios where removing your node_modules directory to correctly install an updated package does not work.

**Comment by marcoscaceres** [comment]
@ianvonholt thanks for the depressing reality check. And yes, having to continually delete and re-install is going to get old very quick - but it does seem to work for a one-off if people get stuck.

**Comment by Mathspy** [comment]
So I have been attempting to reproduce this issue for around an hour now to no avail, although after updating to ``v5.3.0`` I DID hit this issue so I assume that the main cause of this issue is honestly left-over ``package-lock.json``s/``node_module`` folders from pre-``5.3.0``. 

I strongly suggest anyone still facing this issue to completely delete their ``node_modules`` folder AND ``package-lock.json`` and try ``npm install`` again for as many times as they like. Yes the logs will say something silly after a second ``npm install`` like ``added 115 packages`` but that's another issue and it's just a logging issue; ``package-lock.json`` as well as ``node_modules`` remain the same after subsequent ``npm install``s

If you're still able to reproduce this please provide any ``package.json`` for the dev team to use, thank you.

**Comment by jsg2021** [comment]
try turning the lock file off, and have your deps in package.json point to a dist-tag instead of a version/range.

**Comment by ruchernchong** [comment]
@Mathspy Based on what you have mentioned do you do this in the same environment? Because the problem happens when you do `npm install` in development and `npm install` in production.

**Comment by Mathspy** [comment]
@ruchern Tried to fiddle some more with switching between development and production environments but everything seems to be installing as intended, development packages are being installed only in development and folder structure seem to be identical in all situations where the same packages are installed

**Comment by ruchernchong** [comment]
@Mathspy Did any of your npm run scripts break because of missing modules? I always have that problem because the packages were being removed instead.

**Comment by mcollina** [comment]
Typically `eslint` is missing (I depend on standard).

**Comment by Mathspy** [comment]
@ruchern Yes when ``5.3.0`` was first out, my Webpack/Dev-server setup for my current project refused to do anything due to a lot of missing packages but after one completely clean install (like I described above) things started working and I have been installing/uninstalling modules from the project on regular basis without anything breaking so far.
That's why I took a stab at attempting to reproduce this since I have been subscribed to it since it was initially started thinking things will break on next install but they haven't really broken ever since

**Comment by mikemaccana** [comment]
This still occurs in npm 5.4.1 - private packages on the local disk are deleted whenever npm is used. If npm fixes this data-loss bug in future great, but in the meantime, for others having the issue, **if you're using monorepo / modules everywhere**:

 - Install yarn and run `yarn config set workspaces-experimental true`
 - Move your private packages to `/packages`
 - Add a yarn workspace to your `package.json` with:

        "workspaces": [
          "packages/*"
        ],

 - Delete `npm-shrinkwrap.js` and/or `package-lock.json` 
 - Remove any hacks you were doing to restore the packages deleted by npm in your deploy tools
 - Change `npm install` to `yarn` in your deploy tools

**Comment by schmod** [comment]
Just a quick reminder that a lot of folks have commented here saying "This happens to me," and we still do not have documented reproduction steps.

If you want to help resolve this issue, it would be greatly helpful to document the exact set of steps that can reliably demonstrate the bug described in this issue.  

The original post refers to starting with a "dirty" react-native directory structure.  We don't know what the pre-existing `package-lock.json` file or `node_modules` tree looks like.

**Comment by ruchern** [comment]
@schmod If you did read through all the comments, there are even additional steps to reproduce the error on top of the original steps to reproduce in the opening post of this issue. The reason why issues are not being created because it will just get marked as a duplicate of this.

**Comment by schmod** [comment]
Can you please point to them?  The original post refers to starting with a "dirty" react-native directory structure.  We don't know what the pre-existing `package-lock.json` file or `node_modules` tree looks like.

**Comment by mikemaccana** [comment]
@schmod I'm not sure if this and #17927 (which is exclusively regarding `private` modules) are the same issue or different, but I've made a reproduction case for that issue and attached it to the ticket.

**Comment by jbcpollak** [comment]
I just encountered this problem (or a variant) again. 

I manually edited `package.json` with the following diff (along with other changes):

    -    "eslint": "4.5.0",
    +    "eslint": "^4.7.2",

I then ran `npm upgrade`, and `eslint` was missing from `./node_modules`. It was previously installed.

Running `npm install` restored eslint.

**Comment by jokeyrhyme** [comment]
@jbcpollak is that with npm 5.4.2 ? It's been a while, so it's possible this has been addressed?

**Comment by Butterfly** [comment]
Newbie question with eyes on helping: Does `npm` normally delete any dependencies or other packages, like maybe wipe the slate clean, and then replace them with upgraded packages by the same name to meet `npm`'s own upgraded demands?

**Comment by ruchernchong** [comment]
@Butterfly As far as I have used npm v4, npm does not delete package or dependencies. This problem only first appeared when npm went into v5.

**Comment by strugee** [comment]
Note that npm@5 normally does remove extraneous dependencies. The bug here is that it removes virtually _all_ dependencies.

**Comment by iyobo** [comment]
This hasn't been fixed yet huh? I think npm5  is the weakest link for node
v8.x right now and is holding back a lot of adoption of Node 8. instead of having every 'nvm install 8' command immediately followed with 'npm I -g npm@4', perhaps next release of node 8 should just ship with 4 automatically for now until this is sorted out?

**Comment by Butterfly** [comment]
@strugee Thank you for that feedback. I do understand what you're saying.

Why I asked is because I'm wondering if there's even a slim chance that what is happening is that the permissions issue elsewhere could be causing something here, too.

For example, *maybe* `npm` is doing something natural by removing dependencies that it wants to *replace*. But then there's that permissions issue thing going on **elsewhere**.

*Maybe* `npm` is then not able to *put back* a replacement dependency *if* that's even something that it does.

If that was even half a thing going on, it would look like `npm` is only deleting things when *maybe* the problem is that `npm` is not being allowed to complete its intentions of putting a replacement back in place.

Again, I understand you're saying, and this is coming from someone who hasn't grasped exactly what all of `npm` files do yet. I'm saying the above with the thought that removing and replacing same dependencies is likely not part of the scenario. Just never hurts to say things out loud just in case.  :smile: 

A PS is that it would be possible for that to occur and there be no error message, too. I'm trying to help by replicating what others are doing elsewhere and hopefully failing. I did last night. I had to upgrade as `sudo` (Debian/Linux) then I test drove a second upgrade as normal user. It didn't take, it stayed a prior version, AND there was no error message indicating that anything had gone wrong in the process. :smile:

**Comment by jbcpollak** [comment]
In response to @jokeyrhyme 's question:

> @jbcpollak is that with npm 5.4.2 ? It's been a while, so it's possible this has been addressed?

Yes, this was definitely with 5.4.2.

**Comment by jbcpollak** [comment]
Along the lines of what @Butterfly was saying, there is this issue I opened a long time ago:

https://github.com/npm/npm/issues/16834

There are definitely situations where npm@5 will _silently_ fail to install dependencies. The initial issue I found that caused the problem was fixed, but it seems like npm@5 is suppressing installation exceptions that should be reported to the console.

**Comment by strugee** [comment]
@iyobo it hasn't been fixed yet because no one can reliably reproduce it.

@Butterfly right, I see your point. I don't think so though, since npm normally does this kind of stuff with no problem. Plus installing a _new_ dependency should have little to no effect on the existing modules installed - there would be no reason to replace them with a different version.

**Comment by iyobo** [comment]
@strugee I always experience it when I follow these instructions:
http://docs.jollofjs.com/basics/getting-started/

**Comment by strugee** [comment]
@iyobo that's awesome to hear. Maybe now this can get fixed.

What npm and Node are you on?

**Comment by zj-lin-zj** [comment]
I encountered the same issue with npm 5.5.1 and Node 8.9.3 LTS. There doesn't seem be a solution yet but why is this issue closed?

**Comment by BrandonCopley** [comment]
I am having this issue as well, what is the solution?

**Comment by strugee** [comment]
@ACCTFORGH this issue isn't closed. There's no solution because no one has reliably reproduced it (and provided enough details).

**Comment by amejiarosario** [comment]
This is pretty annoying. I have to install everything after adding a dependency. 

```
npm v5.6.0
node v9.3.0
```

**Comment by marcoscaceres** [comment]
Do reproduce, I believe it requires a downgrade to 5.0 and the creation of package-lock.json via npm install. 

To fix (wfm?), delete package-lock.json, make sure you are on latest npm, then npm install again. If necessary, delete your node_modules dir and perform the steps above again.

**Comment by jbcpollak** [comment]
I think @marcoscaceres is correct - this problem seems to have gone away for my team once _everyone_ updated to the latest versions of npm 5. If some people continue to use npm 4 or early npm 5 versions, this problem appears.

**Comment by philjoseph** [comment]
We are all running npm 5.6.0 and some machines expose the problem and some don't. I have no idea what caused it and the `--verbose` is useless to analyze that weird behavior

**Comment by philjoseph** [comment]
Switching over to `yarn` that does expose a more reliable behavior...

**Comment by RoboPhred** [comment]
This may be a duplicate of https://github.com/npm/npm/issues/17379.

The issue seems to happen on any package that does not get an _integrity key on install.  This affects git / github packages and local directory installs, as well as any other install type that pacote does not generate integrity keys for.

**Comment by ghost** [comment]
I'm having the same issue.  npm install of any package on any solution cleans out package-lock making me have to re-install all my packages.  We do use an internal repo, not sure if this is related.

npm - v5.6.0
node - v8.7.0

**Comment by hermione3690** [comment]
same issue
npm 5.6.0
node 8.9.4

**Comment by z-vr** [comment]
`npm` keeps removing git dependencies. It's such a mess. I hope more people switch to `yarn` so that developers take responsibility for their product.

**Comment by erkie** [comment]
@hermione3690 @z-vr If you can reliably reproduce these issues, could you perhaps copy your node_modules, your package.json into a Docker container/git repo and push it up? The problem is that nobody has provided a way for the developers to reproduce this issue on their computers so they can debug what's going on.

**Comment by RoboPhred** [comment]
@erkie I gave a reproduction and cause over at https://github.com/npm/npm/issues/17379#issuecomment-354620986, which is a duplicate issue.

It can be triggered with a package.json containing one git dependency and one non-git dependency:
```
{
  "name": "test-dir",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "my-pkg": "git+http://my-internal-gogs-repo:3000/my-org/my_repo.git#master",
    "lodash.keys": "^4.2.0",
    "lodash.values": "^4.3.0"
  }
}
```

The root cause is that 
 [cloneCurrentTreeToIdealTree](https://github.com/npm/npm/blob/5e426a78ca02d0044f8dd26e0c5f881217081cbd/lib/install.js#L701) (lib/install.js:701) is throwing out packages that lack a _integrity or _shasum field.  Git-installed packages lack this field, because [pacote disables it for git packages](https://github.com/zkat/pacote/blob/latest/lib/fetchers/git.js#L115).

This causes an interesting sequence where inflateShrinkwrap sees the package does not load it into onDisk (because it is not in idealTree), inflatableChild re-downloads the package (because it is not in onDisk), and pruneIdealTree deletes from the tree again (because the node created by inflateShrinkwrap has a userRequired is false).

The net effect is that any ```npm install specific-npm-package``` operation will delete all non-npm git packages, and probably [local directory packages](https://github.com/zkat/pacote/blob/latest/lib/fetchers/directory.js#L31) too.  Annoyingly, it will also completely re-download your git dep while it does this, blocking the install if you do not have access to your git repo.


If we remove the ```_integrity: false``` line from pacote's git repo handling, it happily generates what seems to be a stable hash for the package.  However, we then find that shrinkwrapDeps is [explicitly ignoring git package integrity](https://github.com/npm/npm/blob/latest/lib/shrinkwrap.js#L123), with a comment of:
```
// no integrity for git deps as integirty hashes are based on the
// tarball and we can't (yet) create consistent tarballs from a stable
// source.
```

Removing both makes npm function as I expect; git packages are installed, and remain installed when installing non-git packages.  It also seems that the shrinkwrap comment is wrong and that git hashes are stable, as no reinstalls happen due to integrity changes.  However, I cannot say this for certain, and presumably that comment was there for a reason.  This particular logic was added by @iarna in [shrinkwrap: Don't include integrity for git deps](https://github.com/npm/npm/commit/4baece440b55a981f716fbb1d19e9be13af2520f), and I do not know enough to comment on the cacache content address thing.



At this point, I have a local clone of npm with a fix that has been running stable, but I suspect that I re-introduced whatever that comment was referring to.  Without more feedback, I cannot take this any further.

**Comment by iamtraction** [comment]
So my users and I are starting to face this issues. I've tried to reproduce it and it's happening 100% of the time.

**Steps to reproduce the issue:**
1. Run `npm install`
2. Install another package: `npm install pokedex-api`
3. You'll see a lot of packages will be removed.

Tried with [this `package.json` file](https://github.com/TheBastionBot/Bastion/blob/1ed2f27d9d42a60d788d2188d30493753ece00fc/package.json).

**Reproduction rate:** 100%

**UPDATE:** In step 2, running `npm update` instead of `npm install <package>` has the same effect too.

**Comment by michael-jahn-dev** [comment]
Happens to me also.

`node 8.9.3`
`npm 5.6.0`

I remember that I updated npm a few days ago to 5.6.0 and I think since then I began encountering this issue.

**Comment by felixwahner** [comment]
Same here:
```
~ node -v
v8.9.4
~ npm -v
5.6.0
```

**Comment by VanceHarris** [comment]
Same here
Node 9.0 or 8.9.4
npm 5.6.0

**Comment by strugee** [comment]
All - **please don't leave "me too" comments**. They don't add information and just generate noise for people subscribed to this issue. At this point I believe we have STR so what we really need is for someone to actually fix this bug. Thanks!

**Comment by zontafil** [comment]
@RoboPhred do you think there's a way of adding manually the integrity check in the package-lock.json?

I tried to add the resolved and integrity fields but the package is still getting removed by npm

**Comment by mikemaccana** [comment]
Not advertising yarn here here, but: if this issue is affecting you, you should look consider:

 - how long npm 5 has been stable
 - how much attention this (and other) issues have had from the npm client maintainers

And consciously choose whether you prefer npm5, npm4 or yarn (which have their own tradeoffs).

**Comment by RoboPhred** [comment]
@m3l7 Modifying package-lock will not make a difference here as the package is being thrown out of the install tree due to the package.json lacking the _integrity key, in [install.js:701](https://github.com/npm/npm/blob/5e426a78ca02d0044f8dd26e0c5f881217081cbd/lib/install.js#L701).

Adding an integrity key to your node_module packages one by one might be enough to sustain it, but keep in mind it will disappear next time you update them.

The solution I have been using over the past several weeks is to modify my npm install's code directly to remove the no-integrity purge.

I considered rolling up a pull request as a workaround until pacote figures out the long term solution, but there has been complete silence on this issue so I don't think such a request would even be looked at.

**Comment by z-vr** [comment]
TO EVERYONE HERE!!!

I have found the solution to this problem. What you have to do is 

```
npm i -g yarn
```

and then just use `yarn`. After that, you also get on of the box:

- reliability
- confidence in the next day
- ability to run custom commands only with `yarn custom-command` without having to type `yarn run command` every time, and pass arguments as well: `yarn test --watch`
- speed
- `yarn.lock` which doesn't always change hashes on github
- independence from `npm`
- **ability to have github repositories** specified in `package.json`

! It is that simple. 

Once again, the command is:

```
npm i -g yarn
```

If you feel real geeky, you can do `brew install yarn`, but `npm` is quicker. 

HOPE THIS INFO HELPS!!!

**Comment by joshdick** [comment]
npm 5.7 had [several pruning bugfixes](http://blog.npmjs.org/post/171139955345/v570), maybe it fixes this issue?

**Comment by zkat** [comment]
This is fixed as of `npm@5.7.1`.
